### PR TITLE
SSE at Radosgw end

### PR DESCRIPTION
**Pull request by bafna**
For code review purpose only.

### CODE REVIEW

File: src/common/config_opts.h [hunk 1]
```diff
@@ -965,6 +965,8 @@ OPTION(rgw_swift_auth_entry, OPT_STR, "auth")  // entry point for which a url is
 OPTION(rgw_swift_tenant_name, OPT_STR, "")  // tenant name to use for swift access
 OPTION(rgw_swift_enforce_content_length, OPT_BOOL, false)  // enforce generation of Content-Length even in cost of performance or scalability
 OPTION(rgw_keystone_url, OPT_STR, "")  // url for keystone server
+OPTION(rgw_kms_encrypt_url, OPT_STR, "")  // url for keystone server
+OPTION(rgw_kms_decrypt_url, OPT_STR, "")  // url for keystone server
```
Comment (rahul1aggarwal) [inline src/common/config_opts.h]: can these endpoints be different ? or the API is different ?
Please put proper comment

Comment (bafna) [reply src/common/config_opts.h]: Yes. @harshalgupta gave me two endpoints. Separate for encrypt and decrypt.  Will update the comment.

File: src/rgw/rgw_rest_s3.h [hunk 2]
```diff
@@ -586,6 +586,56 @@ class RGWResourceKeystoneInfo {
...
+    memcpy(ptr, tx_buffer_it.get_current_ptr().c_str(), l);
+    try {
+      tx_buffer_it.advance(l);
+    } catch (buffer::end_of_buffer &e) {
+      assert(0);
+    }
+
+    return l;
+  }
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.h:632]: most of this code is common with RGW_Auth_S3_Keystone_ValidateToken and both these classes are derived from RGWHTTPClient. Can we move this common code to the base class ? these functions are pure virtual in RGWHTTPClient.

File: src/rgw/rgw_rados.cc [hunk 3]
```diff
@@ -9013,3 +9013,131 @@ librados::Rados* RGWRados::get_rados_handle()
...
+
+  return plaintext_len;
+}
+
+
+void handleErrors(void)
+{
+  ERR_print_errors_fp(stderr);
+  abort();
+}
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rados.cc]: we should not abort the process. just fail this particular request.

File: src/rgw/rgw_rest_s3.cc [hunk 4]
```diff
@@ -3239,3 +3310,87 @@ bool RGWResourceKeystoneInfo::get_bucket_public_perm(const string& action,
...
+  {
+    ret = -ERR_INTERNAL_ERROR;
+    return ret; 
+  }
+
+  dout(0) << " gbdebug After parsing " << kmsdata->key_dec << " & " << kmsdata->iv_dec << dendl; 
+
+  kmsdata->key_dec= "0123456789012345678901234567890101234567890123456789012345678901";
+  kmsdata->iv_dec= "01234567890123456";
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc]: remove these hardcoded values

File: src/rgw/rgw_rest_s3.cc [hunk 5]
```diff
@@ -3239,3 +3310,87 @@ bool RGWResourceKeystoneInfo::get_bucket_public_perm(const string& action,
     reason = "OK";
     return true;
 }
+
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc]: please put some comments above every function

File: src/rgw/rgw_rest_s3.cc [hunk 6]
```diff
@@ -3239,3 +3310,87 @@ bool RGWResourceKeystoneInfo::get_bucket_public_perm(const string& action,
...
+  {
+    ret = -ERR_INTERNAL_ERROR;
+    dout(0) << " Unable to obtain encryped and decrypted keys from KMS "<< dendl;
+    return ret; 
+  }
+
+  string bufferprinter = "";
+  rx_buffer.copy(0, rx_buffer.length(), bufferprinter);
+  dout(0) << "gbdebug Printing RX buffer: " << bufferprinter << dendl; 
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc:3330]: log debug messages like it is being done for IAM. also log the time taken

File: src/rgw/rgw_rest_s3.cc [hunk 7]
```diff
@@ -3239,3 +3310,87 @@ bool RGWResourceKeystoneInfo::get_bucket_public_perm(const string& action,
...
     return true;
 }
+
+
+int RGW_KMS::make_kms_encrypt_request(string &root_account, RGWKmsData* kmsdata) 
+{
+  string kms_url = cct->_conf->rgw_kms_encrypt_url;
+  if (kms_url[kms_url.size()] -1 != '/') {
+    kms_url.append("?"); 
+  }
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc:3303]: what is the purpose of this if condition ? 
what is (kms_url[kms_url.size()] -1 != '/') is false ?

File: src/rgw/rgw_rest_s3.cc [hunk 8]
```diff
@@ -3239,3 +3310,87 @@ bool RGWResourceKeystoneInfo::get_bucket_public_perm(const string& action,
...
+  {
+    ret = -ERR_INTERNAL_ERROR;
+    return ret; 
+  }
+
+  dout(0) << " gbdebug After parsing " << kmsdata->key_dec << " & " << kmsdata->iv_dec << dendl; 
+  
+  kmsdata->key_dec = "0123456789012345678901234567890101234567890123456789012345678901";
+  kmsdata->iv_dec = "01234567890123456";
+  // find key,iv by parsing
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc]: remove

File: src/rgw/rgw_op.cc [hunk 9]
```diff
@@ -932,6 +934,22 @@ void RGWGetObj::execute()
   if (ret < 0)
     goto done_err;
 
+  keybl = attrs[RGW_ATTR_KEY];
+  ivbl = attrs[RGW_ATTR_IV];
+  mkeybl = attrs[RGW_ATTR_MKEY];
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_op.cc]: do we get empty bufferlist for attributes which are not present ? or do we need to initialize these bufferlists also ?

File: src/rgw/rgw_op.cc [hunk 10]
```diff
@@ -932,6 +934,22 @@ void RGWGetObj::execute()
...
+    kmsdata = new RGWKmsData();
+    keybl.copy(0,keybl.length(),kmsdata->key_enc);
+    ivbl.copy(0,ivbl.length(),kmsdata->iv_enc);
+    mkeybl.copy(0,mkeybl.length(),kmsdata->mkey_enc);
+    ret = req_kms.make_kms_decrypt_request(root_account,kmsdata);
+    if (ret < 0)
+      return;
+    dout(0) << "gbdebug decrypted "<< dendl;
+    dout(0)  << kmsdata->key_dec.c_str() << dendl;
+    dout(0) << "gbdebug iv" << kmsdata->iv_dec.c_str() << dendl;
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_op.cc]: put proper log messeges. also where is kmsdata being deleted ?

File: src/rgw/rgw_op.cc [hunk 11]
```diff
@@ -1796,14 +1848,34 @@ void RGWPutObj::execute()
   }
 
   processor = select_processor(*(RGWObjectCtx *)s->obj_ctx, &multipart);
-
-  ret = processor->prepare(store, NULL);
+  is_enc = s->info.env->get("HTTP_X_AMZ_SERVER_SIDE_ENCRYPTION");
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_op.cc:1852]: check for HTTP_X_JCS_SERVER_SIDE_ENCRYPTION as well

File: src/rgw/rgw_op.cc [hunk 12]
```diff
@@ -1796,14 +1848,34 @@ void RGWPutObj::execute()
...
+    if (ret < 0)
+      goto done;
+    //Get the key from KMS and store it as attribute
+    dout(0) << "gbdebug : Normal uploading with key " << kmsdata->key_dec << " and " << "iv is " << kmsdata->iv_dec << dendl;
+    keybl.append(kmsdata->key_enc.c_str());
+    ivbl.append(kmsdata->iv_enc.c_str());
+    mkeybl.append(kmsdata->mkey_enc.c_str());
+    attrs[RGW_ATTR_KEY] = keybl;
+    attrs[RGW_ATTR_IV] = ivbl;
+    attrs[RGW_ATTR_MKEY] = mkeybl;
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_op.cc]: should'nt we use master key version here ?

File: src/rgw/rgw_op.cc [hunk 13]
```diff
@@ -2796,6 +2871,29 @@ void RGWInitMultipart::execute()
...
+    dout(0) << "gbdebug : Mutlipart uploading with key " << key << dendl;
+    keybl.append(kmsdata->key_enc.c_str());
+    ivbl.append(kmsdata->iv_enc.c_str());
+    mkeybl.append(kmsdata->mkey_enc.c_str());
+    attrs[RGW_ATTR_KEY] = keybl;
+    attrs[RGW_ATTR_IV] = ivbl;
+    attrs[RGW_ATTR_MKEY] = mkeybl;
+    dout(0) << "gbdebug : Yes enc it is wuth key" << key << dendl;
+  }
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_op.cc:2922]: same code in single part upload. can we make a method for this ?

File: src/rgw/rgw_rados.cc [hunk 14]
```diff
@@ -1089,7 +1089,7 @@ int RGWPutObjProcessor_Atomic::handle_data(bufferlist& bl, off_t ofs, MD5 *hash,
     first_chunk.claim(bl);
     obj_len = (uint64_t)first_chunk.length();
     if (hash) {
-      hash->Update((const byte *)first_chunk.c_str(), obj_len);
+      //hash->Update((const byte *)first_chunk.c_str(), obj_len);
     }
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rados.cc:1093]: not calculating md5 hash anymore ? ok , upadting hash while reading from socket.

File: src/rgw/rgw_rados.cc [hunk 15]
```diff
@@ -1106,7 +1106,7 @@ int RGWPutObjProcessor_Atomic::handle_data(bufferlist& bl, off_t ofs, MD5 *hash,
   int ret = write_data(bl, write_ofs, phandle, exclusive);
   if (ret >= 0) { /* we might return, need to clear bl as it was already sent */
     if (hash) {
-      hash->Update((const byte *)bl.c_str(), bl.length());
+      //hash->Update((const byte *)bl.c_str(), bl.length());
     }
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rados.cc:1110]: not calculating md5 hash anymore ?

File: src/rgw/rgw_rados.cc [hunk 16]
```diff
@@ -9013,3 +9013,131 @@ librados::Rados* RGWRados::get_rados_handle()
...
+{
+  EVP_CIPHER_CTX *ctx;
+
+  int len;
+
+  int ciphertext_len;
+
+  /* Create and initialise the context */
+  if(!(ctx = EVP_CIPHER_CTX_new())) handleErrors();
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rados.cc]: do we need a new context every time ? I am not sure whether this is needed every time or  not and what is the cost

File: src/rgw/rgw_rest.cc [hunk 17]
```diff
@@ -909,7 +909,15 @@ int RGWPutObj_ObjStore::get_params()
...
 
-int RGWPutObj_ObjStore::get_data(bufferlist& bl)
+/* A 512 bit key */
+//unsigned char *key = (unsigned char *)"0123456789012345678901234567890101234567890123456789012345678901";
+
+/* A 128 bit IV */
+//unsigned char *iv = (unsigned char *)"01234567890123456";
+
+
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest.cc]: remove

File: src/rgw/rgw_rest.cc [hunk 18]
```diff
@@ -922,15 +930,92 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
...
+          read_data[i] = '0'; //for now
+        read_len = 32; 
+      }
+      len = read_len;
+
+      /* Initialise the library */
+      ERR_load_crypto_strings();
+      OpenSSL_add_all_algorithms();
+      OPENSSL_config(NULL);
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest.cc]: do we need this everytime ?

File: src/rgw/rgw_rest.cc [hunk 19]
```diff
@@ -922,15 +930,92 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
...
+        return r;
+      unsigned char* read_data = reinterpret_cast<unsigned char *>(bp.c_str());
+      if (hash && read_len)
+        hash->Update((const byte *)bp.c_str(),read_len);
+
+      dout(0) << "gbdebug Earlier Read length " << read_len << " " << kmsdata->key_dec << " is" << dendl;
+      if (read_len < 16 && read_len > 0)
+      {
+        unsigned deficit = 16 - read_len;
+        s->content_length += deficit;
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest.cc]: why is this needed ?

File: src/rgw/rgw_rest.cc [hunk 20]
```diff
@@ -922,15 +930,92 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
...
+      }
+      else if (read_len == 16 && read_len > 0)
+      {
+        s->content_length += 16;
+        for (int i = read_len; i < 32; i++)
+          read_data[i] = '0'; //for now
+        read_len = 32; 
+      }
+      len = read_len;
+
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest.cc:959]: why are you padding the data ?

File: src/rgw/rgw_rest.cc [hunk 21]
```diff
@@ -922,15 +930,92 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
...
+      /* Initialise the library */
+      ERR_load_crypto_strings();
+      OpenSSL_add_all_algorithms();
+      OPENSSL_config(NULL);
+
+      unsigned char ciphertext[2*read_len];
+      //unsigned char decryptedtext[2*read_len];
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest.cc]: why 2*read_len ? Also, dont allocate on the stack. too many put requests can lead to stack overflow.

File: src/rgw/rgw_rest.cc [hunk 22]
```diff
@@ -922,15 +930,92 @@ int RGWPutObj_ObjStore::get_data(bufferlist& bl)
...
+        hash->Update((const byte *)bp.c_str(),read_len);
+
+      dout(0) << "gbdebug Earlier Read length " << read_len << " " << kmsdata->key_dec << " is" << dendl;
+      if (read_len < 16 && read_len > 0)
+      {
+        unsigned deficit = 16 - read_len;
+        s->content_length += deficit;
+        dout(0) << "gbdebug deficit " << deficit << " " << dendl;
+        for (int i = read_len ; i < 16 ; i++)
+          read_data[i] = deficit + '0'; //for now
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest.cc]: is this correct ?

File: src/rgw/rgw_op.cc (hunk 9 above)
Comment (bafna) [reply src/rgw/rgw_op.cc]: We get an empty bufferlist for those.

File: src/rgw/rgw_rest_s3.cc [hunk 23]
```diff
@@ -72,17 +73,83 @@ static struct response_attr_param resp_attr_params[] = {
...
-  bufferlist metadata_bl;
+  bufferlist metadata_bl, decrypted_bl;
+  if (kmsdata)
+  {
+    unsigned char* read_data; 
+    char right_part[16]; 
+    int decryptedtext_len,left_data,iter, full_chunks ;
+    char padded[] = "0000000000000000";
+    uint64_t chunk_size = s->cct->_conf->rgw_max_chunk_size;
+    unsigned char decryptedtext[chunk_size];
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc]: not on stack

File: src/rgw/rgw_rest_s3.cc [hunk 24]
```diff
@@ -72,17 +73,83 @@ static struct response_attr_param resp_attr_params[] = {
...
+    else if (left_data == 32)
+    {
+      strncpy(right_part,(char*)decryptedtext + 16, (unsigned)16); 
+      if (!(strncmp(right_part,padded,16)))
+      {
+        decrypted_bl.append((char*)decryptedtext,16);
+        bl_len = bl_len - 16;
+        total_len = total_len - 16;
+        dout(0) << "gbdebug Exact 16 bytes" << bl_len << dendl;
+      }
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc:121]: what if my original data was of length k*(chunk_size) + 32 (with all zeroes)

File: src/rgw/rgw_rest_s3.cc [hunk 25]
```diff
@@ -72,17 +73,83 @@ static struct response_attr_param resp_attr_params[] = {
...
+      else
+        decrypted_bl.append((char*)decryptedtext,32);
+      decryptedtext[decryptedtext_len] = '\0';
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc]: why null terminating ? for logging ?

File: src/rgw/rgw_rest_s3.cc [hunk 26]
```diff
@@ -72,17 +73,83 @@ static struct response_attr_param resp_attr_params[] = {
...
+    {
+      deficit_char = decryptedtext[15];
+      unsigned deficit = (deficit_char - '0');
+      bl_len = bl_len - deficit;
+      total_len = total_len - deficit;
+      dout(0) << "gbdebug Not Exact 16 bytes" << deficit << "Deficit" << "And" << bl_len << dendl;
+      decrypted_bl.append((char*)decryptedtext,16 - deficit);
+      decryptedtext[decryptedtext_len] = '\0';
+      dout(0) << "gbdebug Real Decrypted text " << decryptedtext << dendl;
+    }
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_rest_s3.cc]: what is left_data > 16 && left_data < 32

File: src/rgw/rgw_rest_s3.h (hunk 2 above)
Comment (bafna) [reply src/rgw/rgw_rest_s3.h:632]: RGWRESTSimpleRequest is also using RGWHTTPClient and having different definitions for them .

File: src/rgw/rgw_main.cc [hunk 27]
```diff
@@ -1241,7 +1241,11 @@ int main(int argc, const char **argv)
 
   /* setup openssl multi-threading functions */
   openssl_thread_setup();
-  
+  /* Initialise the library */
+  ERR_load_crypto_strings();
+  OpenSSL_add_all_algorithms();
+  OPENSSL_config(NULL);
+  srand ( time(NULL) );
```
Comment (rahul1aggarwal) [inline src/rgw/rgw_main.cc:1248]: is this random number initialization needed for openssl library ?

Comment (bafna) [reply src/rgw/rgw_main.cc:1248]: This is for random id we are appending for KMS Debugging. Only for staging. Will remove in prod

**Pull request merged**